hyperspace/cosmos/src/provider.rs

```diff
@@ -0,0 +1,1127 @@
+use super::{
+	client::CosmosClient,
+	events::{
+		event_is_type_channel, event_is_type_client, event_is_type_connection,
+		ibc_event_try_from_abci_event, IbcEventWithHeight,
+	},
+};
+use crate::error::Error;
+use futures::{
+	stream::{self, select_all},
+	Stream, StreamExt,
+};
+use ibc::{
+	applications::transfer::{Amount, BaseDenom, PrefixedCoin, PrefixedDenom, TracePath},
+	core::{
+		ics02_client::{
+			client_state::ClientType, events as ClientEvents,
+			msgs::update_client::MsgUpdateAnyClient, trust_threshold::TrustThreshold,
+		},
+		ics04_channel::packet::Sequence,
+		ics23_commitment::{commitment::CommitmentPrefix, specs::ProofSpecs},
+		ics24_host::{
+			identifier::{ChainId, ChannelId, ClientId, ConnectionId, PortId},
+			path::{
+				AcksPath, ChannelEndsPath, ClientConsensusStatePath, ClientStatePath,
+				CommitmentsPath, ConnectionsPath, Path, ReceiptsPath, SeqRecvsPath,
+			},
+		},
+	},
+	events::IbcEvent,
+	protobuf::Protobuf,
+	timestamp::Timestamp,
+	tx_msg::Msg,
+	Height,
+};
+use ibc_primitives::PacketInfo as IbcPacketInfo;
+use ibc_proto::{
+	cosmos::bank::v1beta1::QueryBalanceRequest,
+	google::protobuf::Any,
+	ibc::core::{
+		channel::v1::{
+			Channel, QueryChannelResponse, QueryChannelsRequest, QueryChannelsResponse,
+			QueryConnectionChannelsRequest, QueryNextSequenceReceiveResponse,
+			QueryPacketAcknowledgementResponse, QueryPacketAcknowledgementsRequest,
+			QueryPacketCommitmentResponse, QueryPacketCommitmentsRequest,
+			QueryPacketReceiptResponse, QueryUnreceivedAcksRequest, QueryUnreceivedPacketsRequest,
+		},
+		client::v1::{
+			QueryClientStateResponse, QueryClientStatesRequest, QueryConsensusStateResponse,
+		},
+		connection::v1::{
+			ConnectionEnd, IdentifiedConnection, QueryConnectionResponse, QueryConnectionsRequest,
+		},
+	},
+};
+use ibc_rpc::PacketInfo;
+use ics07_tendermint::{
+	client_message::ClientMessage, client_state::ClientState, consensus_state::ConsensusState,
+};
+use pallet_ibc::light_clients::{
+	AnyClientMessage, AnyClientState, AnyConsensusState, HostFunctionsManager,
+};
+use primitives::{mock::LocalClientTypes, Chain, IbcProvider, UpdateType};
+use prost::Message;
+use std::{pin::Pin, str::FromStr, time::Duration};
+use tendermint::block::Height as TmHeight;
+pub use tendermint::Hash;
+use tendermint_rpc::{
+	endpoint::tx::Response,
+	event::{Event, EventData},
+	query::{EventType, Query},
+	Client, Error as RpcError, Order, SubscriptionClient, WebSocketClient,
+};
+use tokio::time::sleep;
+
+#[derive(Clone, Debug)]
+pub enum FinalityEvent {
+	Tendermint { from: TmHeight, to: TmHeight },
+}
+
+#[derive(Clone, Debug)]
+pub struct TransactionId<Hash> {
+	pub hash: Hash,
+}
+
+#[async_trait::async_trait]
+impl<H> IbcProvider for CosmosClient<H>
+where
+	H: Clone + Send + Sync + 'static,
+{
+	type FinalityEvent = FinalityEvent;
+	type TransactionId = TransactionId<Hash>;
+	type AssetId = String;
+	type Error = Error;
+
+	async fn query_latest_ibc_events<C>(
```

Comment:
can you break this function into sub functions ?So that it's easier to follow what's goin' on. I find it a bit too long.

Reply:
Done